### ISSUE TITLE
Cask URL stats — broken URLs, insecure redirects

### ISSUE THREAD

**Issue by vszakats** [Homebrew/homebrew-cask#52886]
Out of 7980 unique HTTPS Cask URLs, 269 (3.3%) were inaccessible due to errors and 66 (0.8%) were redirected to an insecure URL. The test didn't check for _intermediate_ insecure redirects, only the landing URL, but intermediate insecure links with a secure link at the end are very rare anyway. Of the 66 URLs, 33% were `homepage`, rest is `url`/`appcast`.

Out of the 269 failures some may be attributed to the way the test was done which may be different from how Cask would do the download, but from Homebrew-core experience, this only affects a fraction of percent of all the cases.

Half of the failures is 404 or 403, rest is various server configuration issues, timeouts and HTTPS configuration errors — IOW, general URL-rot.

A portion of insecure redirects still do support HTTPS, but the redirect is wrongly configured and keeps pointing to an insecure URL. These be fixed manually or by auto-correcting automatic redirects in multiple iterations.

Takeaways from these stats:

- Insecure redirects already do happen with Cask URLs
  (These may or may not be detected at the time of committing them to the repository.)

- Insecure redirects _are rare_, both relative to the total number of URLs and also to the number of actual failures.

- These tests also took homepages into account, even though homepages are opened interactively in browsers by humans, where any redirect (or mixed-content and other potential issue) is readily visible so they do no harm or infer a false sense of security. The benefit of using the secure alternative here, is that the page will automatically be fixed right away once the server configuration is updated and/or the page fixed (e.g. when a SourceForge project admin enabled HTTPS for a given user page.)

These stats also indicate that if upgrading all potential 850 Cask URLs to HTTPS, the amount of extra insecure redirects would fall in the ballpark of 7 URLs possibly having such issues. Blocking the whole
upgrade effort to avoid those ~7 cases looks like the wrong thing to do, though after doing the upgrade, my quick-and-dirty stat script can be re-run and those 7 fixed along with the existing 66 cases, if that's deemed critical. (There exist an extra 75 potential upgrades for `project.sourceforge.net`-style homepages, where most of them redirect to HTTP, unless their maintainer flip a switch to use HTTPS. Notice that upgrade is automatically enforced for all non-Cask formulae since long. It'd also be nice to align URL expectations across Homebrew Taps.)

The current state of Casks URLs lag about 2 years behind that of Homebrew-core, where Casks use 78% HTTPS vs. insecure URLs, whereas Homebrew-core is at [94%](https://github.com/Homebrew/homebrew-core/pull/32692#issuecomment-427302530).

<details><summary>Click for raw results.</summary><p>

```
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://airvpn.org/macosx/ -> http://airvpn.org/macos/
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 500 
curl: (22) The requested URL returned error: 403 Forbidden
redir to insecure: https://assets.unigine.com/d/Unigine_Heaven-4.0.dmg -> http://m12-assets.unigine.com/d/Unigine_Heaven-4.0.dmg
redir to insecure: https://assets.unigine.com/d/Unigine_Valley-1.0.dmg -> http://m11-assets.unigine.com/d/Unigine_Valley-1.0.dmg
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
redir to insecure: https://basecamp.com/help/ -> http://basecamp.com/help
redir to insecure: https://beamdog.com/about/client -> http://www.beamdog.com/about/client
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://binarynights.com/forklift/ -> http://binarynights.com/
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (7) Failed to connect to cache-ams04.cdn.yandex.net port 443: Connection refused
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 500 Internal Server Error
curl: (28) Connection time-out
redir to insecure: https://chazmcgarvey.github.com/thecheat/thecheat-1.2.5.dmg -> http://chazmcgarvey.github.io/thecheat/thecheat-1.2.5.dmg
curl: (7) Failed to connect to cliqz.com port 443: Connection refused
curl: (22) The requested URL returned error: 404 Not Found
curl: (28) Operation timed out after 10000 milliseconds with 0 bytes received
curl: (22) The requested URL returned error: 403 
curl: (7) Failed to connect to crisp.chat port 443: Connection refused
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
redir to insecure: https://deepstream.io/ -> http://deepstreamhub.com/open-source/?io
redir to insecure: https://denysdovhan.com/inboxer -> http://denysdovhan.com/inboxer/
curl: (22) The requested URL returned error: 501 Not Implemented
curl: (7) Failed to connect to devrantron.firebaseapp.com port 443: Connection refused
curl: (22) The requested URL returned error: 500 
curl: (22) The requested URL returned error: 401 
curl: (22) The requested URL returned error: 405 Method Not Allowed
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (7) Failed to connect to download.cdn.yandex.net port 443: Connection refused
curl: (7) Failed to connect to download.cdn.yandex.net port 443: Connection refused
redir to insecure: https://download.kde.org/stable/digikam/digiKam-5.9.0-01-MacOS-x86-64.pkg -> http://mirrors.ocf.berkeley.edu/kde/stable/digikam/digiKam-5.9.0-01-MacOS-x86-64.pkg
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 404 Not Found
curl: (52) Empty reply from server
curl: (6) Could not resolve host: edgehill.nylas.com
curl: (60) SSL certificate problem: self signed certificate
curl: (22) The requested URL returned error: 500 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 404 
redir to insecure: https://fliqlo.com/download/fliqlo_171.dmg -> http://fliqlo.com/
curl: (6) Could not resolve host: gateblu.octoblu.com
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 404 
redir to insecure: https://glassfish.java.net/docs/4.0/installation-guide.pdf -> http://javaee.github.io/glassfish
curl: (22) The requested URL returned error: 404 
redir to insecure: https://hbang.ws/apps/termhere -> http://hbang.ws/apps/termhere/
curl: (22) The requested URL returned error: 405 /
curl: (51) SSL: no alternative certificate subject name matches target host name 'igrsoft.com'
curl: (28) SSL connection timeout
curl: (22) The requested URL returned error: 403 
redir to insecure: https://imazing.com/mini -> http://imazing.com/backup-iphone-ipad
curl: (22) The requested URL returned error: 525 
curl: (7) Failed to connect to jalbum.net port 443: Connection refused
curl: (51) SSL: no alternative certificate subject name matches target host name 'kompozer.net'
redir to insecure: https://leagueoflegends.com/ -> http://www.leagueoflegends.com/
redir to insecure: https://library.panic.com/releasenotes/transmit -> http://library.panic.com/releasenotes/transmit/
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 402 
curl: (22) The requested URL returned error: 500 Internal Server Error
curl: (22) The requested URL returned error: 405 Method Not Allowed
curl: (22) The requested URL returned error: 501 Not Implemented
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (60) SSL certificate problem: certificate has expired
curl: (60) SSL certificate problem: certificate has expired
redir to insecure: https://mbdapp.iqiyi.com/j/ot/iQIYIMedia_005.dmg -> http://cdn.data.video.iqiyi.com/cdn/pcclient/20180830/19/02/iQIYIMedia_005.dmg
curl: (28) Operation timed out after 10002 milliseconds with 0 bytes received
curl: (28) Operation timed out after 10002 milliseconds with 0 bytes received
curl: (28) Connection timed out after 5003 milliseconds
redir to insecure: https://messengerfordesktop.org/ -> http://www.messengerfordesktop.org/
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://mongobooster.com/ -> http://nosqlbooster.com/
curl: (22) The requested URL returned error: 405 Method Not Allowed
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://panic.com/candybar/d/CandyBar%203.3.4.zip -> http://download.panic.com/candybar/CandyBar%203.3.4.zip
curl: (22) The requested URL returned error: 405 Method Not Allowed
curl: (7) Failed to connect to pc.115.com port 443: Connection refused
curl: (22) The requested URL returned error: 404 File Not Found
curl: (51) SSL: no alternative certificate subject name matches target host name 'pictures.lytro.com'
curl: (22) The requested URL returned error: 404 Not Found
curl: (51) SSL: no alternative certificate subject name matches target host name 'pki.io'
curl: (7) Failed to connect to play.google.com port 443: Connection refused
redir to insecure: https://plot.ly/free-sql-client-download -> http://plot.ly/free-sql-client-download/
curl: (22) The requested URL returned error: 503 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
curl: (28) Connection timed out after 5000 milliseconds
curl: (22) The requested URL returned error: 405 Method Not Allowed
curl: (22) The requested URL returned error: 403 
curl: (51) SSL: no alternative certificate subject name matches target host name 'robofont.com'
curl: (6) Could not resolve host: royaltsx-v
curl: (22) The requested URL returned error: 503 
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 404 Not Found
curl: (56) LibreSSL SSL_read: SSL_ERROR_SYSCALL, errno 54
redir to insecure: https://sites.box.com/sync4/ -> http://www.box.com
curl: (22) The requested URL returned error: 404 Not Found
curl: (60) SSL certificate problem: unable to get local issuer certificate
curl: (60) SSL certificate problem: unable to get local issuer certificate
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 404 
redir to insecure: https://sparkler.herokuapp.com/apps/3/versions/132/download -> http://s3.amazonaws.com/sparkler_versions/versions/uploads/000/000/132/original/Anvil.zip?1513340725
curl: (22) The requested URL returned error: 405 METHOD NOT ALLOWED
curl: (22) The requested URL returned error: 404 
redir to insecure: https://srf.baidu.com/?c=j&e=d&platform=mac -> http://imeres.baidu.com/mac/baiduinput_mac_v5.2.0.22_1000e.dmg
curl: (28) Operation timed out after 10002 milliseconds with 0 bytes received
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (7) Failed to connect to storage.crisp.chat port 443: Connection refused
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (22) The requested URL returned error: 401 Unauthorized
curl: (22) The requested URL returned error: 401 Unauthorized
curl: (22) The requested URL returned error: 401 Unauthorized
redir to insecure: https://support.apple.com/downloads/DL1433/en_US/RecoveryDiskAssistant.dmg -> http://download.info.apple.com/Mac_OS_X/041-2001.20110808.QFC1d/RecoveryDiskAssistant.dmg
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
curl: (28) Connection timed out after 5000 milliseconds
redir to insecure: https://thoughtbot.github.io/sassquatch/download/v1.5.0/Sassquatch.zip -> http://sassquatch.thoughtbot.com/download/v1.5.0/Sassquatch.zip
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Appcast not found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 500 Internal Server Error
curl: (7) Failed to connect to vk.com port 443: Connection refused
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
redir to insecure: https://www.5kplayer.com/download/5kplayer.dmg -> http://dl1.5kplayer.com/download/5kplayer.dmg
curl: (28) Connection timed out after 5003 milliseconds
curl: (22) The requested URL returned error: 401 
curl: (22) The requested URL returned error: 401 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 Not Found
^[[A^[[Acurl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://www.bitgapp.com/eqmac/ -> http://bitgapp.com/eqmac/
curl: (28) Connection timed out after 5000 milliseconds
curl: (28) Connection timed out after 5001 milliseconds
curl: (22) The requested URL returned error: 406 Not Acceptable
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 405 
redir to insecure: https://www.colwiz.com/ -> http://www.colwiz.com/landing
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://www.deezer.com/desktop/download/artifact/darwin/x64/0.14.4 -> http://e-cdn-content.deezer.com/builds/deezer-desktop/8cF2rAuKxLcU1oMDmCYm8Uiqe19Ql0HTySLssdzLkQ9ZWHuDTp2JBtQOvdrFzWPA/darwin/x64/0.14.4/DeezerDesktop_0.14.4.dmg
redir to insecure: https://www.deezer.com/formac -> http://www.deezer.com/download
curl: (51) SSL: no alternative certificate subject name matches target host name 'www.delightfuldev.com'
redir to insecure: https://www.digitalrebellion.com/download/cineplay?version=1440 -> http://download.digitalrebellion.com/cineplay/CinePlay_1440.dmg
redir to insecure: https://www.digitalrebellion.com/download/fcsremover?version=3140 -> http://download.digitalrebellion.com/fcsremover/FCS_Remover_3140.dmg
redir to insecure: https://www.digitalrebellion.com/download/posthaste?version=2630 -> http://download.digitalrebellion.com/posthaste/Post_Haste_for_Mac_2630.dmg
redir to insecure: https://www.dssw.co.uk/authbuddy/dsswauthbuddy.dmg -> http://www.dssw.co.uk/authbuddy/dsswauthbuddy-110.dmg
redir to insecure: https://www.dssw.co.uk/sleepmonitor/dsswsleepmonitor.dmg -> http://www.dssw.co.uk/sleepmonitor/dsswsleepmonitor-302.dmg
redir to insecure: https://www.earslap.com/assets/laconicism/Laconicism.dmg -> http://earslap.com/assets/laconicism/Laconicism.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/modeling/mdt/papyrus/rcp/oxygen/3.0.0/papyrus-oxygen-3.0.0-macosx64.tar.gz&r=1 -> http://download.eclipse.org/modeling/mdt/papyrus/rcp/oxygen/3.0.0/papyrus-oxygen-3.0.0-macosx64.tar.gz
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/2018-09/R/eclipse-jee-2018-09-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/2018-09/R/eclipse-jee-2018-09-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-dsl-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-dsl-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-javascript-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-javascript-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-modeling-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-modeling-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-parallel-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-parallel-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-php-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-php-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-rcp-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-rcp-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-scout-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-scout-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/oxygen/3a/eclipse-testing-oxygen-3a-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/oxygen/3a/eclipse-testing-oxygen-3a-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/photon/R/eclipse-committers-photon-R-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/photon/R/eclipse-committers-photon-R-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/photon/R/eclipse-cpp-photon-R-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/photon/R/eclipse-cpp-photon-R-macosx-cocoa-x86_64.dmg
redir to insecure: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/photon/R/eclipse-java-photon-R-macosx-cocoa-x86_64.dmg&r=1 -> http://download.eclipse.org/technology/epp/downloads/release/photon/R/eclipse-java-photon-R-macosx-cocoa-x86_64.dmg
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 405 METHOD NOT ALLOWED
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 404 Not Found
curl: (28) Operation timed out after 5002 milliseconds with 0 out of 0 bytes received
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (7) Failed to connect to www.hidemyass.com port 443: Connection refused
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 404 
curl: (60) SSL certificate problem: certificate has expired
redir to insecure: https://www.imobie.com/product/anytrans-mac.dmg -> http://dl.imobie.com/anytrans-mac.dmg
curl: (60) SSL certificate problem: certificate has expired
curl: (60) SSL certificate problem: certificate has expired
curl: (60) SSL certificate problem: certificate has expired
curl: (60) SSL certificate problem: certificate has expired
curl: (60) SSL certificate problem: certificate has expired
curl: (22) The requested URL returned error: 501 Not Implemented
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (22) The requested URL returned error: 403 Forbidden
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (92) HTTP/2 stream 1 was not closed cleanly: PROTOCOL_ERROR (err 1)
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://www.macupdate.com/download/34101/SafariSort.dmg -> http://dl.macupdate.com/prod/SafariSort.dmg
curl: (22) The requested URL returned error: 404 Page Not Found
curl: (22) The requested URL returned error: 503 
curl: (22) The requested URL returned error: 503 
curl: (28) Operation timed out after 10001 milliseconds with 0 bytes received
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 405 
curl: (22) The requested URL returned error: 405 
redir to insecure: https://www.osomac.com/appcasts/handbrakebatch/HandBrakeBatch.xml -> http://osomac.com/appcasts/handbrakebatch/HandBrakeBatch.xml
redir to insecure: https://www.osomac.com/appcasts/handbrakebatch/HandBrakeBatch_2.25.zip -> http://osomac.com/appcasts/handbrakebatch/HandBrakeBatch_2.25.zip
redir to insecure: https://www.osomac.com/apps/osx/handbrake-batch/ -> http://osomac.com/apps/osx/handbrake-batch/
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 
curl: (60) SSL certificate problem: certificate has expired
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 405 Method Not Allowed
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 404 
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 403 
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 500 Internal Server Error
redir to insecure: https://www.shinywhitebox.com/store/file.php?p=1 -> http://www.shinywhitebox.com/downloads/iShowU_1.94.5.dmg
redir to insecure: https://www.shinywhitebox.com/store/file.php?p=12 -> http://www.shinywhitebox.com/downloads/iShowU_Studio_1.7.6.dmg
redir to insecure: https://www.shinywhitebox.com/store/file.php?p=14 -> http://www.shinywhitebox.com/downloads/instant/iShowU_Instant_1.2.8.dmg
redir to insecure: https://www.shinywhitebox.com/store/file.php?p=5 -> http://www.shinywhitebox.com/downloads/iShowU_HD_2.3.21.dmg
curl: (22) The requested URL returned error: 404 
curl: (35) error:14004438:SSL routines:CONNECT_CR_SRVR_HELLO:tlsv1 alert internal error
redir to insecure: https://www.subclassed.com/download/Mnemosyne.zip -> http://www.subclassed.com/files/sw/Mnemosyne.zip
curl: (22) The requested URL returned error: 403 
curl: (7) Failed to connect to www.sugarsync.com port 443: Connection refused
curl: (7) Failed to connect to www.sugarsync.com port 443: Connection refused
curl: (60) SSL certificate problem: unable to get local issuer certificate
curl: (22) The requested URL returned error: 404 
curl: (35) error:1400510A:SSL routines:CONNECT_CR_SRVR_HELLO:wrong ssl version
curl: (35) error:1400510A:SSL routines:CONNECT_CR_SRVR_HELLO:wrong ssl version
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://www.timeedition.com/ -> http://www.timeedition.com/en/
curl: (22) The requested URL returned error: 501 Not Implemented
redir to insecure: https://www.tuxera.com/mac/tuxerantfs_2018.dmg -> http://download.tuxera.com/mac/tuxerantfs_2018.dmg
redir to insecure: https://www.varaneckas.com/jad/jad158g.mac.intel.zip -> http://varaneckas.com/jad/jad158g.mac.intel.zip
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://www.waterfoxproject.org/ -> http://www.waterfoxproject.org/en-US/
curl: (22) The requested URL returned error: 404 Not Found
redir to insecure: https://www.youneedabudget.com/CDNOrigin/download/ynab4/liveCaptive/Mac/YNAB4_LiveCaptive_4.3.855.dmg -> http://classic.youneedabudget.com//CDNOrigin/download/ynab4/liveCaptive/Mac/YNAB4_LiveCaptive_4.3.855.dmg
curl: (28) Resolving timed out after 5540 milliseconds
curl: (28) Resolving timed out after 5542 milliseconds
curl: (28) Resolving timed out after 5548 milliseconds
curl: (22) The requested URL returned error: 404 Not Found
curl: (22) The requested URL returned error: 564 
curl: (7) Failed to connect to yandex.ru port 443: Connection refused
redir to insecure: https://yllan.github.com/nally/download/Nally-1.4.9.app.zip -> http://yllan.github.io/nally/download/Nally-1.4.9.app.zip
curl: (6) Could not resolve host: zcash4mac.com
curl: (28) Operation timed out after 5002 milliseconds with 0 out of 0 bytes received
```
</p></details>

**Comment by vitorgalvao** [comment]
Why not post this in the PR? You were the one who closed it and I already said I agree with you. If you care to reopen it we’ll continue the discussion there, and possibly even merge it.

Thank you for the stats. Closing since the issue is non-actionable and you already have the PR to fix this.

---

Small point, though.

> The current state of Casks URLs lag about 2 years behind that of Homebrew-core

Not a fair comparison at all. HB only has open-source URLs, which are typically hosted in platforms that do it for them (such as GitHub). HBC takes URLs from anywhere, and we have no control over if the developers use HTTP or HTTPS.

**Comment by vszakats** [comment]
Sorry, I deemed the PR abandoned and started these tests in the meanwhile.

I agree it's not 1-to-1 comparison between Casks and Formulae, but the high number of low-hanging upgrade options show the room for potential. As an interesting data-point, about _half_ of Cask HTTPS URLs are `https://github.com/` ones, so GitHub's share is significant (I omitted them from these tests as they are known to be working fine.)

One more thought I wanted to share here: My testing method only considered redirects via the `Location:` header. For homepages there are other ways to do it, which are all problematic to detect with automated tools.

...Let's continue on the PR, though I'll have to create a new one to re-upload the patch.
